overlord/ifacestate/handlers.go

```diff
@@ -485,6 +487,143 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+// doAutoConnect creates task(s) to connect the given snap to viable candidates.
+func (m *InterfaceManager) doAutoConnect(task *state.Task, _ *tomb.Tomb) error {
+	// FIXME: here we should not reconnect auto-connect plug/slot
+	// pairs that were explicitly disconnected by the user
+
+	var conns map[string]connState
+	st := task.State()
+	autochecker, err := newAutoConnectChecker(task.State())
+	if err != nil {
+		return err
+	}
+
+	st.Lock()
+	defer st.Unlock()
+
+	err = st.Get("conns", &conns)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	if conns == nil {
+		conns = make(map[string]connState)
+	}
+
+	snapsup, err := snapstate.TaskSnapSetup(task)
+	if err != nil {
+		return err
+	}
+
+	snapName := snapsup.Name()
+
+	var autoconnectTs []*state.TaskSet
+
+	// Auto-connect all the plugs
+	for _, plug := range m.repo.Plugs(snapName) {
+		candidates := m.repo.AutoConnectCandidateSlots(snapName, plug.Name, autochecker.check)
+		if len(candidates) == 0 {
+			continue
+		}
+		// If we are in a core transition we may have both the old ubuntu-core
+		// snap and the new core snap providing the same interface. In that
+		// situation we want to ignore any candidates in ubuntu-core and simply
+		// go with those from the new core snap.
+		if len(candidates) == 2 {
+			switch {
+			case candidates[0].Snap.Name() == "ubuntu-core" && candidates[1].Snap.Name() == "core":
+				candidates = candidates[1:2]
+			case candidates[1].Snap.Name() == "ubuntu-core" && candidates[0].Snap.Name() == "core":
+				candidates = candidates[0:1]
+			}
+		}
+		if len(candidates) != 1 {
+			crefs := make([]string, 0, len(candidates))
+			for _, candidate := range candidates {
+				crefs = append(crefs, candidate.String())
+			}
+			task.Logf("cannot auto connect %s (plug auto-connection), candidates found: %q", plug, strings.Join(crefs, ", "))
```

Comment:
_"cannot auto-connect plug %s, candidates found: %s"_

Note the %s at the end as well.. it feels awkward to read a list that is quoted as if it were a single element.